File: mkdocs/snippets/devbook/accounts/configure_multisig.mjs
import { PrivateKey } from 'symbol-sdk';
import {
	NemFacade,
	calculateTransactionFee,
	descriptors,
	models
} from 'symbol-sdk/nem';

const NODE_URL = process.env.NODE_URL ||
	'http://libertalia.nemtest.net:7890';
console.log('Using node', NODE_URL);

const facade = new NemFacade('testnet');
// [>step-1]
const KEY_PREFIX = '0'.repeat(63);

// Set up the keys for the multisig account and its two cosignatories
const MULTISIG_PRIVATE_KEY = process.env.MULTISIG_PRIVATE_KEY || (
	`${KEY_PREFIX}1`);
const multisigKeyPair = new NemFacade.KeyPair(
	new PrivateKey(MULTISIG_PRIVATE_KEY));
const multisigAddress = facade.network.publicKeyToAddress(
	multisigKeyPair.publicKey);
console.log(`Multisig address: ${multisigAddress}`,
	`(public key ${multisigKeyPair.publicKey})`);

const cosignatoryKeyPairs = [];
for (let i = 0; 2 > i; i++) {
	const COSIGNATORY_PRIVATE_KEY =
		process.env[`COSIGNATORY${i}_PRIVATE_KEY`] || (
			KEY_PREFIX + String(i + 2));
	const keyPair = new NemFacade.KeyPair(
		new PrivateKey(COSIGNATORY_PRIVATE_KEY));
	cosignatoryKeyPairs.push(keyPair);
	const addr = facade.network.publicKeyToAddress(keyPair.publicKey);
	console.log(`Cosignatory ${i} address: ${addr}`,
		`(public key ${keyPair.publicKey})`);
}
// [<step-1]
// Helper function to announce a transaction
async function announceTransaction(payload, label) {
	const announcePath = '/transaction/announce';
	console.log(`Announcing ${label} to ${announcePath}`);
	const response = await fetch(`${NODE_URL}${announcePath}`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: payload
	});
	if (!response.ok)
		throw new Error(`HTTP ${response.status}`);
	const result = await response.json();
	console.log('  Result:', result.message);
	return result.message;
}

// Helper function to wait for transaction confirmation
async function waitForConfirmation(transactionHash, label) {
	const statusPath = `/transaction/get?hash=${transactionHash}`;
	console.log(`Waiting for ${label} confirmation from`, statusPath);
	let isConfirmed = false;
	for (let attempt = 1; 120 >= attempt; ++attempt) {
		const response = await fetch(`${NODE_URL}${statusPath}`);
		if (!response.ok) {
			console.log('  Transaction status: pending');
			await new Promise(resolve => { setTimeout(resolve, 1000); });
		} else {
			const confirmed = await response.json();
			console.log(`${label} confirmed in block`,
				confirmed.meta.height);
			isConfirmed = true;
			break;
		}
	}
	if (!isConfirmed)
		console.warn(`${label} confirmation took too long.`);
}

// Returns the cosignatory addresses of the provided multisig [>step-2]
// account, or an empty list if the account is not multisig
async function getMultisigCosignatories(address) {
	const accountPath = `/account/get?address=${address}`;
	console.log(`Getting cosignatories from ${accountPath}`);
	const response = await fetch(`${NODE_URL}${accountPath}`);
	const accountInfo = await response.json();
	const foundCosignatories = accountInfo.meta.cosignatories
		.map(cosignatory => cosignatory.address);
	if (0 === foundCosignatories.length) {
		console.log('  Response: No cosignatories');
		return [];
	}
	console.log('  Response:', JSON.stringify(foundCosignatories));
	return foundCosignatories;
}
// [<step-2]
// [>step-4]
// Returns a transaction that turns a regular account into a multisig
function multisigEnableTransaction(approvalDelta) {
	// Create a multisig account modification transaction
	// that adds the cosignatories
	const modifications = cosignatoryKeyPairs.map(keyPair =>
		new descriptors.SizePrefixedMultisigAccountModificationDescriptor(
			new descriptors.MultisigAccountModificationDescriptor(
				models.MultisigAccountModificationType.ADD_COSIGNATORY,
				keyPair.publicKey)));
	const transaction = facade.createTransactionFromTypedDescriptor(
		new descriptors.MultisigAccountModificationTransactionV2Descriptor(
		// Change of the number of cosignatures
		// required to approve transactions
			approvalDelta,
			modifications),
		// This is the account that will be turned into a multisig
		multisigKeyPair.publicKey,
		0n,
		2 * 60 * 60);
	// [<step-4]
	// Calculate and attach the transaction fee [>step-5]
	const fee = calculateTransactionFee(transaction);
	transaction.fee = new models.Amount(fee);
	console.log(`  Transaction fee: ${Number(fee) / 1_000_000} XEM`);
	console.log(
		'Enabling the multisig with the modification transaction:');
	console.log(JSON.stringify(transaction.toJson(), null, 2));
	// [<step-5]
	// Sign the transaction with the multisig's key [>step-6]
	const signature = facade.signTransaction(
		multisigKeyPair, transaction);
	facade.transactionFactory.static.attachSignature(
		transaction, signature);
	return transaction; // [<step-6]
}

// [>step-7]
// Returns a transaction that removes one cosignatory from the multisig
function multisigRemovalTransaction(removedKeyPair, approvalDelta) {
	// Create a multisig account modification transaction
	// that removes a single cosignatory
	const innerTransaction = facade.createTransactionFromTypedDescriptor(
		new descriptors.MultisigAccountModificationTransactionV2Descriptor(
		// Change of the number of cosignatures
		// required to approve transactions
			approvalDelta,
			[new descriptors.SizePrefixedMultisigAccountModificationDescriptor(
				new descriptors.MultisigAccountModificationDescriptor(
					models.MultisigAccountModificationType.DELETE_COSIGNATORY,
					removedKeyPair.publicKey))]),
		// This is the multisig account that will be modified
		multisigKeyPair.publicKey,
		0n,
		2 * 60 * 60);
	// [<step-7]
	// Wrap the modification in a multisig transaction [>step-8]
	const innerFee = calculateTransactionFee(innerTransaction);
	innerTransaction.fee = new models.Amount(innerFee);
	const transaction = facade.createTransactionFromTypedDescriptor(
		new descriptors.MultisigTransactionV1Descriptor(
			facade.transactionFactory.static.toNonVerifiableTransaction(
				innerTransaction)),
		// This is the cosignatory that initiates the removal
		cosignatoryKeyPairs[0].publicKey,
		0n,
		2 * 60 * 60);
	// [<step-8]
	// Calculate and attach the transaction fee [>step-9]
	const fee = calculateTransactionFee(transaction);
	transaction.fee = new models.Amount(fee);
	console.log('  Transaction fee:',
		`${Number(innerFee + fee) / 1_000_000} XEM`);
	console.log(
		'Disabling the multisig with the multisig transaction:');
	console.log(JSON.stringify(transaction.toJson(), null, 2));
	// [<step-9]
	// Sign the transaction with the cosignatory's key [>step-10]
	const signature = facade.signTransaction(
		cosignatoryKeyPairs[0], transaction);
	facade.transactionFactory.static
		.attachSignature(transaction, signature);
	return transaction; // [<step-10]
}

try {
	// Get current state of the multisig account and decide [>step-3]
	// which operation to perform
	const cosignatories = await getMultisigCosignatories(multisigAddress);
	let transactions;
	if (0 === cosignatories.length) {
		// Enable the multisig
		transactions = [multisigEnableTransaction(1)];
	} else {
		// Disable the multisig
		transactions = [
			multisigRemovalTransaction(cosignatoryKeyPairs[1], 0),
			multisigRemovalTransaction(cosignatoryKeyPairs[0], -1)
		];
	}
	// [<step-3]
	// Announce each transaction and wait for confirmation [>step-11]
	for (const signedTransaction of transactions) {
		const transactionHash = facade.hashTransaction(signedTransaction)
			.toString();
		console.log('Built transaction with hash:', transactionHash);
		const jsonPayload = facade.transactionFactory.static
			.toJson(signedTransaction);
		const result = await announceTransaction(
			jsonPayload, 'transaction');
		if ('SUCCESS' !== result) {
			console.log('Transaction rejected');
			break;
		}
		await waitForConfirmation(transactionHash, 'transaction');
	}
	// [<step-11]
} catch (e) {
	console.error(e.message, '| Cause:', e.cause?.code ?? 'unknown');
}

File: mkdocs/snippets/devbook/accounts/configure_multisig.py
import json
import os
import time
import urllib.request

from symbolchain.CryptoTypes import PrivateKey
from symbolchain.facade.NemFacade import NemFacade
from symbolchain.nc import Amount
from symbolchain.nem.FeeCalculator import calculate_transaction_fee
from symbolchain.nem.Network import NetworkTimestamp

NODE_URL = os.getenv('NODE_URL', 'http://libertalia.nemtest.net:7890')
print(f'Using node {NODE_URL}')

facade = NemFacade('testnet')
# [>step-1]
KEY_TEMPLATE = '0' * 63 + '{}'

# Set up the keys for the multisig account and its two cosignatories
MULTISIG_PRIVATE_KEY = os.getenv(
	'MULTISIG_PRIVATE_KEY', KEY_TEMPLATE.format(1))
multisig_key_pair = NemFacade.KeyPair(PrivateKey(MULTISIG_PRIVATE_KEY))
multisig_address = facade.network.public_key_to_address(
	multisig_key_pair.public_key)
print(f'Multisig address: {multisig_address} '
	f'(public key {multisig_key_pair.public_key})')

cosignatory_key_pairs = []
for i in range(2):
	COSIGNATORY_PRIVATE_KEY = os.getenv(
		f'COSIGNATORY{i}_PRIVATE_KEY', KEY_TEMPLATE.format(i + 2))
	key_pair = NemFacade.KeyPair(PrivateKey(COSIGNATORY_PRIVATE_KEY))
	cosignatory_key_pairs.append(key_pair)
	addr = facade.network.public_key_to_address(key_pair.public_key)
	print(f'Cosignatory {i} address: '
		f'{addr} (public key {key_pair.public_key})')  # [<step-1]


# Helper function to announce a transaction
def announce_transaction(payload, label):
	announce_path = '/transaction/announce'
	print(f'Announcing {label} to {announce_path}')
	request = urllib.request.Request(
		f'{NODE_URL}{announce_path}',
		data=payload.encode(),
		headers={'Content-Type': 'application/json'},
		method='POST'
	)
	with urllib.request.urlopen(request) as announce_response:
		result = json.loads(announce_response.read().decode())
	print(f'  Result: {result["message"]}')
	return result['message']


# Helper function to wait for transaction confirmation
def wait_for_confirmation(tx_hash, label):
	status_path = f'/transaction/get?hash={tx_hash}'
	print(f'Waiting for {label} confirmation from {status_path}')
	is_confirmed = False
	for _ in range(120):
		try:
			with urllib.request.urlopen(
				f'{NODE_URL}{status_path}'
			) as status_response:
				confirmed = json.loads(status_response.read().decode())
				height = confirmed['meta']['height']
				print(f'{label} confirmed in block {height}')
				is_confirmed = True
				break
		except urllib.error.HTTPError:
			print('  Transaction status: pending')
		time.sleep(1)
	if not is_confirmed:
		print(f'{label} confirmation took too long.')


# Returns the cosignatory addresses of the provided multisig [>step-3]
# account, or an empty list if the account is not multisig
def get_multisig_cosignatories(address):
	account_path = f'/account/get?address={address}'
	print(f'Getting cosignatories from {account_path}')
	url = f'{NODE_URL}{account_path}'
	with urllib.request.urlopen(url) as account_response:
		account_info = json.loads(account_response.read().decode())
		found_cosignatories = [
			cosignatory['address']
			for cosignatory in account_info['meta']['cosignatories']
		]
		if not found_cosignatories:
			print('  Response: No cosignatories')
			return []
		print(f'  Response: {found_cosignatories}')
		return found_cosignatories  # [<step-3]


# [>step-5]
# Returns a transaction that turns a regular account into a multisig
def multisig_enable_transaction(tx_timestamp, tx_deadline,
		approval_delta):
	# Create a multisig account modification transaction
	# that adds the cosignatories
	modifications = [
		{'modification': {
			'modification_type': 'add_cosignatory',
			'cosignatory_public_key': key_pair.public_key
		}}
		for key_pair in cosignatory_key_pairs
	]
	transaction = facade.transaction_factory.create({
		'type': 'multisig_account_modification_transaction_v2',
		# This is the account that will be turned into a multisig
		'signer_public_key': multisig_key_pair.public_key,
		'timestamp': tx_timestamp.timestamp,
		'deadline': tx_deadline.timestamp,
		# Change of the number of cosignatures
		# required to approve transactions
		'min_approval_delta': approval_delta,
		'modifications': modifications
	})
	# [<step-5]
	# Calculate and attach the transaction fee [>step-6]
	fee = calculate_transaction_fee(transaction)
	transaction.fee = Amount(fee)
	print(f'  Transaction fee: {fee / 1_000_000} XEM')
	print('Enabling the multisig with the modification transaction:')
	print(json.dumps(transaction.to_json(), indent=2))
	# [<step-6]
	# Sign the transaction with the multisig's key [>step-7]
	signature = facade.sign_transaction(multisig_key_pair, transaction)
	facade.transaction_factory.attach_signature(transaction, signature)
	return transaction  # [<step-7]


# [>step-8]
# Returns a transaction that removes one cosignatory from the multisig
def multisig_removal_transaction(tx_timestamp, tx_deadline,
		removed_key_pair, approval_delta):
	# Create a multisig account modification transaction
	# that removes a single cosignatory
	inner_transaction = facade.transaction_factory.create({
		'type': 'multisig_account_modification_transaction_v2',
		# This is the multisig account that will be modified
		'signer_public_key': multisig_key_pair.public_key,
		'timestamp': tx_timestamp.timestamp,
		'deadline': tx_deadline.timestamp,
		# Change of the number of cosignatures
		# required to approve transactions
		'min_approval_delta': approval_delta,
		'modifications': [
			{'modification': {
				'modification_type': 'delete_cosignatory',
				'cosignatory_public_key': removed_key_pair.public_key
			}}
		]
	})
	# [<step-8]
	# Wrap the modification in a multisig transaction [>step-9]
	inner_fee = calculate_transaction_fee(inner_transaction)
	inner_transaction.fee = Amount(inner_fee)
	transaction = facade.transaction_factory.create({
		'type': 'multisig_transaction_v1',
		# This is the cosignatory that initiates the removal
		'signer_public_key': cosignatory_key_pairs[0].public_key,
		'timestamp': tx_timestamp.timestamp,
		'deadline': tx_deadline.timestamp,
		'inner_transaction':
			facade.transaction_factory.to_non_verifiable_transaction(
				inner_transaction)
	})
	# [<step-9]
	# Calculate and attach the transaction fee [>step-10]
	fee = calculate_transaction_fee(transaction)
	transaction.fee = Amount(fee)
	print(f'  Transaction fee: {(inner_fee + fee) / 1_000_000} XEM')
	print('Disabling the multisig with the multisig transaction:')
	print(json.dumps(transaction.to_json(), indent=2))
	# [<step-10]
	# Sign the transaction with the cosignatory's key [>step-11]
	signature = facade.sign_transaction(
		cosignatory_key_pairs[0], transaction)
	facade.transaction_factory.attach_signature(transaction, signature)
	return transaction  # [<step-11]


try:
	# Fetch current network time [>step-2]
	time_path = '/time-sync/network-time'
	print(f'Fetching current network time from {time_path}')
	with urllib.request.urlopen(f'{NODE_URL}{time_path}') as response:
		response_json = json.loads(response.read().decode())
		network_time = response_json['receiveTimeStamp'] // 1000
		print(f'  Network time: {network_time} s since the nemesis block')

	# Derived fields from network time
	timestamp = NetworkTimestamp(network_time)
	deadline = timestamp.add_hours(2)
	# [<step-2]
	# Get current state of the multisig account and decide which [>step-4]
	# operation to perform
	cosignatories = get_multisig_cosignatories(multisig_address)
	if len(cosignatories) == 0:
		# Enable the multisig
		transactions = [multisig_enable_transaction(
			timestamp, deadline, 1)]
	else:
		# Disable the multisig
		transactions = [
			multisig_removal_transaction(
				timestamp, deadline, cosignatory_key_pairs[1], 0),
			multisig_removal_transaction(
				timestamp, deadline, cosignatory_key_pairs[0], -1)
		]
	# [<step-4]
	# Announce each transaction and wait for confirmation [>step-12]
	for signed_transaction in transactions:
		transaction_hash = facade.hash_transaction(signed_transaction)
		print(f'Built transaction with hash: {transaction_hash}')
		json_payload = facade.transaction_factory.to_json(
			signed_transaction)
		announce_result = announce_transaction(
			json_payload, 'transaction')
		if 'SUCCESS' != announce_result:
			print('Transaction rejected')
			break
		wait_for_confirmation(transaction_hash, 'transaction')
	# [<step-12]
except Exception as e:
	print(e)
